src/lib/ClaimProcessorFunctionCastLib.sol

```diff
@@ -0,0 +1,935 @@
+
+import {
+    BatchClaim,
+    QualifiedBatchClaim,
+    BatchClaimWithWitness,
+    QualifiedBatchClaimWithWitness,
+    SplitBatchClaim,
+    SplitBatchClaimWithWitness,
+    QualifiedSplitBatchClaim,
+    QualifiedSplitBatchClaimWithWitness
+} from "../types/BatchClaims.sol";
+
+import {
+    MultichainClaim,
+    QualifiedMultichainClaim,
+    MultichainClaimWithWitness,
+    QualifiedMultichainClaimWithWitness,
+    SplitMultichainClaim,
+    SplitMultichainClaimWithWitness,
+    QualifiedSplitMultichainClaim,
+    QualifiedSplitMultichainClaimWithWitness,
+    ExogenousMultichainClaim,
+    ExogenousQualifiedMultichainClaim,
+    ExogenousMultichainClaimWithWitness,
+    ExogenousQualifiedMultichainClaimWithWitness,
+    ExogenousSplitMultichainClaim,
+    ExogenousSplitMultichainClaimWithWitness,
+    ExogenousQualifiedSplitMultichainClaim,
+    ExogenousQualifiedSplitMultichainClaimWithWitness
+} from "../types/MultichainClaims.sol";
+
+import {
+    BatchMultichainClaim,
+    QualifiedBatchMultichainClaim,
+    BatchMultichainClaimWithWitness,
+    QualifiedBatchMultichainClaimWithWitness,
+    SplitBatchMultichainClaim,
+    SplitBatchMultichainClaimWithWitness,
+    QualifiedSplitBatchMultichainClaim,
+    QualifiedSplitBatchMultichainClaimWithWitness,
+    ExogenousBatchMultichainClaim,
+    ExogenousQualifiedBatchMultichainClaim,
+    ExogenousBatchMultichainClaimWithWitness,
+    ExogenousQualifiedBatchMultichainClaimWithWitness,
+    ExogenousSplitBatchMultichainClaim,
+    ExogenousSplitBatchMultichainClaimWithWitness,
+    ExogenousQualifiedSplitBatchMultichainClaim,
+    ExogenousQualifiedSplitBatchMultichainClaimWithWitness
+} from "../types/BatchMultichainClaims.sol";
+
+/**
+ * @title ClaimProcessorFunctionCastLib
+ * @notice Libray contract implementing function casts used in ClaimProcessorLogic.
+ * The input function operates on a function that takes some argument that differs
+ * from what is currently available. The output function modifies one or more
+ * argument types so that they match the arguments that are being used to call the
+ * function. Note that from the perspective of the function being modified, the
+ * original type is still in force; great care should be taken to preserve offsets
+ * and general structure between the two structs.
+ * @dev Note that some of these function casts may no longer be in use.
+ */
+library ClaimProcessorFunctionCastLib {
+    /**
+     * @notice Function cast to provide a BasicClaim calldata struct while
+     * treating it as a uint256 representing a calldata pointer location.
+     * @param fnIn   Function pointer to `ClaimProcessorLib.processSimpleClaim`.
+     * @return fnOut Modified function used in `ClaimProcessorLogic._processBasicClaim`.
+     */
+    function usingBasicClaim(function(bytes32, uint256, uint256, bytes32, bytes32, function(address, address, uint256, uint256) internal returns (bool)) internal returns (bool) fnIn)
```

Comment:
Can you explain at a high level why this is necessary? I've read the natspec but still not sure. Which one is the correct function type for the function thats going to be called? Why would the type ever be incorrect and need changing?

Reply:
Yeah, using this function as an example:
- `ClaimProcessorLib` has a function `processSimpleClaim` that is used as a shared entrypoint for processing four different claim types (basic claim, claim with witness, multichain claim, and multichain claim with witness) where the only key differences are how the claim hash is derived and what the calldata offset to the relevant data is. To facilitate this, the `processSimpleClaim` function takes a `uint256 calldataPointer`  argument instead of like a `BasicClaim` calldata argument.
- When we're actually _calling_ `processSimpleClaim` inside of `ClaimProcessorLogic._processBasicClaim`, the `claimPayload` has  a `BasicClaim calldata claimPayload` type. While you _could_ do a typecast to a uint256 inside of that function before calling `processSimpleClaim`, it uses extra gas and critically takes up more contract size, so instead we can use a function cast as a "zero-cost abstraction" to accomplish the same thing
- The `fnIn` function matches the arguments that `processSimpleClaim` expects, but the `fnOut` function matches the argument types that are actually being provided